### PR TITLE
Liming

### PR DESCRIPTION
**Pull request by liming-vie**
Add table factory for create tables and smart allocator for Blob

### CODE REVIEW

File: src/util/allocator.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  data = (char*)_aligned_malloc(size, \
+    MV_CONFIG_allocator_alignment);     
+#define ALIGN_FREE(data)              \
+  _aligned_free(data);
+#else   
+#define ALIGN_MALLOC(data, size)               \
+  CHECK(posix_memalign(&data,                  \
+    MV_CONFIG_allocator_alignment, size) == 0);
+#define ALIGN_FREE(data)    free(data);
+#endif
```
Comment (feiga) [inline src/util/allocator.cpp]: using function instead of Macro

File: include/multiverso/util/allocator.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+private:
+  char* data_;
+  std::atomic<int> ref_ = 0;
+  static const size_t header_size_ = (sizeof(MemoryBlock*) << 1);
+};
+
+class Allocator {
+public:
+  virtual ~Allocator() = default;
+  virtual char* Malloc(size_t size);
```
Comment (feiga) [inline include/multiverso/util/allocator.h]: Alloc, instead of malloc

File: include/multiverso/blob.h [hunk 3]
```diff
@@ -4,67 +4,83 @@
...
+      memcpy(data_, data, size_);
+    }
+
+    Blob(void* data, size_t size) : size_(size) {
+      data_ = Allocator::Get()->Malloc(size);
+      memcpy(data_, data, size_);
+    }
+
+    Blob(const Blob& rhs) {
+      Allocator::Get()->Refer(rhs.data_);
```
Comment (feiga) [inline include/multiverso/blob.h]: make reference count in Blob, instead of memory block from allocator

File: src/table/array_table.cpp [hunk 4]
```diff
@@ -19,14 +19,17 @@ ArrayWorker<T>::ArrayWorker(size_t size) : WorkerTable(), size_(size) {
   Log::Debug("worker %d create arrayTable with %d elements.\n", MV_Rank(), size);
 }
 
+template <typename T>
+ArrayWorker<T>::ArrayWorker(const ArrayTableInitOption &option) {
+  new (this)ArrayWorker(option.size);
```
Comment (feiga) [inline src/table/array_table.cpp]: it's strange. 

I think the following should be fine. 

ArrayWorker<T>::ArrayWorker(const ArrayTableOption &option) : ArrayWorker(option.size) {
}

File: src/table/array_table.cpp [hunk 5]
```diff
@@ -84,6 +87,11 @@ ArrayServer<T>::ArrayServer(size_t size) : ServerTable() {
 }
 
 template <typename T>
+ArrayServer<T>::ArrayServer(const ArrayTableInitOption& option) {
+  new (this)ArrayServer(option.size);
```
Comment (feiga) [inline src/table/array_table.cpp]: same with worker

File: include/multiverso/table/array_table.h [hunk 6]
```diff
@@ -7,11 +7,15 @@
 
 namespace multiverso {
 
+struct ArrayTableInitOption {
```
Comment (feiga) [inline include/multiverso/table/array_table.h]: ArrayTableOption, remove Init

File: include/multiverso/multiverso.h [hunk 7]
```diff
@@ -23,6 +24,16 @@ int  MV_ServerId();
...
 int  MV_ServerIdToRank(int server_id);
 
+// create server table and worker table
+// \param option for table initiate
+// \return worker table pointer if this node is worker
+//  otherwise return nullptr
+template <typename EleType, typename OptionType>
+typename trait::OptionTrait<EleType, OptionType>::WorkerTableType*
+MV_CreateTable(const OptionType& option) {
+  return table_factory::CreateTable<EleType>(option);
```
Comment (feiga) [inline include/multiverso/multiverso.h]: Add a Barrier here? Then user will not call Barrier any more.

**Comment by chivee** [comment]
hi  @liming-vie , this PR make the master can't pass the build in Linux, please check the issue.

**Pull request merged**